go/merkle/encoder.go

```diff
@@ -0,0 +1,71 @@
+package merkle
+
+import (
+	"crypto/hmac"
+	cryptorand "crypto/rand"
+	"crypto/sha256"
+	"crypto/sha512"
+
+	"github.com/pkg/errors"
+)
+
+type Encoder struct {
+	encodingType EncodingType
+}
+
+func NewEncoder(encodingType EncodingType) *Encoder {
+	return &Encoder{encodingType: encodingType}
+}
+
+func (e *Encoder) BlindedPreimage(leaf Leaf, key Key, secret Secret) (BlindedPreimage, error) {
+	switch e.encodingType {
+	case EncodingTypeBlindedSHA256v1:
+		h := hmac.New(sha256.New, secret.Secret)
+		h.Write(key.Key)
+		return NewBlindedPreimage(leaf, h.Sum(nil))
+	case EncodingTypeBlindedSHA512_256v1:
+		h := hmac.New(sha512.New512_256, secret.Secret)
+		h.Write(key.Key)
+		return NewBlindedPreimage(leaf, h.Sum(nil))
+	default:
+		return BlindedPreimage{}, errors.Errorf("unknown encoding type %q", e.encodingType)
+	}
+}
+
+func (e *Encoder) Hash(preimage BlindedPreimage) ([]byte, error) {
+	b, err := preimage.LeafContainer.Serialize()
+	if err != nil {
+		return nil, err
+	}
+	switch e.encodingType {
+	case EncodingTypeBlindedSHA256v1:
```

Comment:
why do we need this?

Reply:
no reason in particular, can get rid of it

Reply:
👍 yeah let's keep it down to as few options and features as possible